OrbitBase/ThreadPool.cpp

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "OrbitBase/ThreadPool.h"
+
+#include <list>
+#include <thread>
+
+#include "OrbitBase/Logging.h"
+#include "absl/synchronization/mutex.h"
+
+namespace {
+
+class ThreadPoolImpl : public ThreadPool {
+ public:
+  explicit ThreadPoolImpl(size_t thread_pool_min_size,
+                          size_t thread_pool_max_size, size_t queue_threshold);
+
+  void Schedule(std::unique_ptr<Action> action) override;
+  void Shutdown() override;
+  void Wait() override;
+
+ private:
+  bool ActionsAvailableOrShutdownInitiated();
+  // Blocking call - returns nullptr if shutdown initiated.
+  std::unique_ptr<Action> PopAction();
+  void CreateWorker();
+  void WorkerFunction();
+
+  absl::Mutex mutex_;
+  std::list<std::unique_ptr<Action>> scheduled_actions_;
+  std::vector<std::thread> worker_threads_;
+  size_t thread_pool_max_size_;
+  size_t queue_threshold_;
+  bool shutdown_initiated_;
+};
+
+ThreadPoolImpl::ThreadPoolImpl(size_t thread_pool_min_size,
+                               size_t thread_pool_max_size,
+                               size_t queue_threshold)
+    : thread_pool_max_size_(thread_pool_max_size),
+      queue_threshold_(queue_threshold),
+      shutdown_initiated_(false) {
+  CHECK(thread_pool_min_size > 0);
+  CHECK(thread_pool_max_size >= thread_pool_min_size);
+  CHECK(queue_threshold > 0);
```

Comment:
If min_size == max_size, then queue_threshold is meaningless (right?), so I would naively assume that creating a ThreadPool with ThreadPool(8, 8, 0) should be fine, but would cause an abort here. 

Maybe this condition should be : CHECK(thread_pool_min_size == thread_pool_max_size || queue_threshold > 0);

Reply:
agree - done.